Upcoming elections homepage CTA

symroe opened the pull request:
I thought a table was easier to read than a list, and it looks like this:

![screenshot-2018-5-30 democracy club candidates](https://user-images.githubusercontent.com/242329/40736837-03829bae-6437-11e8-8f42-8c22f1f37909.png)

I don't think it does a great job of asking / telling people what to do, but for those who know it will help. Do you think we should try to explain things more?

symroe commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Upcoming by-elections on YNR](https://trello.com/c/eGjbPjpJ/1334-upcoming-by-elections-on-ynr)

chris48s commented:
In general this is really useful *content* :+1:

I think it does a good job of providing a summary to engaged/regular users and also addresses issue https://github.com/DemocracyClub/yournextrepresentative/issues/446

We should definitely present this somewhere on the site, but I don't think it makes good front-page content. There's 2 reasons for this:

1. As you note in the PR comment

> I don't think it does a great job of asking / telling people what to do

Basically, this (also, think about https://github.com/DemocracyClub/yournextrepresentative/issues/484 ). The first thing users front page should be orientated towards explaining what this is, or providing one easy CTA (e.g: sign up, find your local area). Starting off with a big grid and non-obvious terminology like "locked" isn't a good place to start IMO. Maybe if we start thinking about 'levels' or 'profiles', this is good front page content once you are a user who has made 50+ edits, but not for an unauthenticated user the first time they hit the site.

2. Mobile compatibility

I think its correct that a table is a good way to represent this info, but have a look at this page in responsive design mode. I reckon we should try and make sure that if the first time someone hits the site is on a phone, at least the front page looks good. Horizontal scroll bars etc are going to be a turn-off esp for first-timers.

chris48s commented on ynr/apps/elections/uk/templatetags/home_page_tags.py:
```diff
@@ -141,3 +141,27 @@ def results_progress(context):
+                        takes_context=True)
+def by_election_ctas(context):
+    context['SHOW_BY_ELECTION_CTA'] = getattr(
+        settings, 'SHOW_BY_ELECTION_CTA', False)
```
Do you think we should set a default value for this in `local.py.example` in this PR?

Also, in the case where this is `False`, can we hide the `<div>` completely so that we don't show an empty box:

![screenshot at 2018-05-30 20 45 25](https://user-images.githubusercontent.com/6025893/40744514-a9b9ffb6-644c-11e8-9f42-5d9e80c875ce.png)

symroe replied:
This isn't a new problem in this PR – there are a few things that could be in that box, and basically one of them always is. Currently there is `SHOW_ELECTION_STATS`,  `SHOW_SOPN_TRACKER`,  `SHOW_RESULTS_PROGRESS` and  `SHOW_BY_ELECTION_CTA`. Any any point, one of these will be `True`, but there isn't a process for ensuring that exactly one is True at the moment.

What do you think a good way of doing this would be?

chris48s replied:
Maybe another way to look at it would be that if we always want exactly one thing to be true, then we don't want several boolean settings, we want one setting (e.g: `FRONTPAGE_CTA`) which is a string like `sopn_import_progress` or `current_election_stats` that tells us what template to include? If that setting is missing/invalid, we don't show anything.

Given this is unrelated to the changes in this PR, happy to address as a different issue to avoid scope creep in this PR.

symroe replied:
Right, that's nice. Happy to make that change in this PR as it's not especially important that this is merged quickly, so we may as well pay off some debt while we're here and talking about it.

chris48s commented on ynr/apps/elections/uk/templates/includes/by-election-ctas.html:
```diff
@@ -0,0 +1,73 @@
+        </tr>
+        </thead>
+        <tbody>
+            {% for pee in upcoming_pees %}
```
I expect we're unlikely to hit this very often in real-world usage, but this could be made slightly more elegant by adding a case that hides whole table/headers in the case where `len(upcoming_pees) == 0` (and similarly below if `len(past_pees) == 0` ) instead of rendering a table with 0 rows.

symroe commented:
So as an idea, I've removed the last couple of columns if the user isn't logged in. 

This addresses the unclear CTA issue and mobile issue for new users. 

For existing users, it's not true that everyone who is authenticated will know what to do with the table, but until we have some sort of threshold system it might be good enough?

As for mobile performance: this is one site where I never bother checking, mainly because I assume that doing anything the site asks you to do on a mobile would be quite painful. I don't know how much we should worry about this, but I do agree that the home page at least should be ok to look at on a mobile.

chris48s commented:
> until we have some sort of threshold system it might be good enough?

Yeah, lets not try and boil the ocean here. Logged in/not logged in is a reasonable level of granularity for the moment..

> As for mobile performance: this is one site where I never bother checking, mainly because I assume that doing anything the site asks you to do on a mobile would be quite painful. I don't know how much we should worry about this, but I do agree that the home page at least should be ok to look at on a mobile.

Yes - I think we're completely agreed on this :+1: 

If you're trying to copy and paste info out of PDFs on your phone you are having a bad time and we shouldn't be trying to optimise for that.

Simultaneously, I'm also conscious that there will be times when someone links to the site (e.g: https://medium.com/@meandvan/crowdsourced-factchecking-4c5168ea5ac3 ) and some people reading it will go "what is this Democracy Club Candidates?" and click the link. If they're on a phone, it would be nice if the thing they hit looks reasonable and makes some level of sense.

symroe commented:
I've now updated the order so that the by-elections is under the 'default' CTA: 

![screenshot-2018-5-31 democracy club candidates](https://user-images.githubusercontent.com/242329/40779625-9219f470-64cd-11e8-91e0-13a46044b370.png)

It makes the home page longer, but it also gives the impression of 'currentness' (e.g, that we cover all elections). It also doesn't make the CTA worse, IMHO, but we have no idea of how good that is. As Joe says, this is something for actual research in future.

In non-visual changes, I've moved to using a `FRONT_PAGE_CTA` setting as discussed above.
